### PR TITLE
Port swift/Immediate to Windows

### PR DESCRIPTION
**Pull request by hughbe**
- Fix _MSC_VER condition to _WIN32
- Add WIN32_LEAN_AND_MEAN and NOMINMAX
- Fix LoadRuntimeLib to work when compiling with unicode
- Change use of dlsym that is not available on Windows

### CODE REVIEW

File: lib/Immediate/Immediate.cpp [hunk 1]
```diff
@@ -39,18 +39,21 @@
...
 using namespace swift;
 using namespace swift::immediate;
 
-static void *loadRuntimeLib(StringRef runtimeLibPathWithName) {
-#if defined(_MSC_VER)
-  return LoadLibrary(runtimeLibPathWithName.str().c_str());
+static void* loadRuntimeLib(StringRef runtimeLibPathWithName) {
+#if defined(_WIN32)
+  auto x = LoadLibraryA(runtimeLibPathWithName.str().c_str());
+  return (void*)x;
```
Comment (compnerd) [inline lib/Immediate/Immediate.cpp]: Please use a C++ style cast.

File: lib/Immediate/Immediate.cpp [hunk 2]
```diff
@@ -319,10 +322,17 @@ int swift::RunImmediately(CompilerInstance &CI, const ProcessCmdLine &CmdLine,
 
   // Setup interpreted process arguments.
   using ArgOverride = void (*)(const char **, int);
+  #if defined(_WIN32)
+  auto emplaceProcessArgs
+          = (ArgOverride)GetProcAddress(static_cast<HMODULE>(stdlib), "_swift_stdlib_overrideUnsafeArgvArgc");
```
Comment (compnerd) [inline lib/Immediate/Immediate.cpp]: Please use a C++ style cast.  clang-format the line.  It looks pretty long.

**Review by compnerd — CHANGES_REQUESTED** [review]

File: lib/Immediate/Immediate.cpp [hunk 3]
```diff
@@ -39,18 +39,21 @@
...
 #include <dlfcn.h>
 #endif
 
 using namespace swift;
 using namespace swift::immediate;
 
-static void *loadRuntimeLib(StringRef runtimeLibPathWithName) {
-#if defined(_MSC_VER)
-  return LoadLibrary(runtimeLibPathWithName.str().c_str());
+static void* loadRuntimeLib(StringRef runtimeLibPathWithName) {
```
Comment (CodaFi) [inline lib/Immediate/Immediate.cpp]: Nit: Why did the pointer star move?

File: lib/Immediate/Immediate.cpp (hunk 2 above)
Comment (hughbe) [inline lib/Immediate/Immediate.cpp]: done and formatted

File: lib/Immediate/Immediate.cpp (hunk 1 above)
Comment (hughbe) [inline lib/Immediate/Immediate.cpp]: done, thanks

File: lib/Immediate/Immediate.cpp (hunk 3 above)
Comment (hughbe) [inline lib/Immediate/Immediate.cpp]: accidental, fixed

File: lib/Immediate/Immediate.cpp [hunk 4]
```diff
@@ -319,10 +321,18 @@ int swift::RunImmediately(CompilerInstance &CI, const ProcessCmdLine &CmdLine,
 
   // Setup interpreted process arguments.
   using ArgOverride = void (*)(const char **, int);
+  #if defined(_WIN32)
```
Comment (jrose-apple) [inline lib/Immediate/Immediate.cpp]: Nitpick: Please leave all preprocessor directives unindented.

Comment (hughbe) [inline lib/Immediate/Immediate.cpp]: fixed, my bad

**Comment by hughbe** [comment]
@jrose-apple could you trigger testing for this. I assume I should also apply for testing privileges right?

**Comment by CodaFi** [comment]
@hughbe I got you.

@swift-ci please smoke test.

**Comment by jrose-apple** [comment]
LGTM when @compnerd's happy with it.

**Comment by hughbe** [comment]
@compnerd PTAL a when you have time, thanks

**Review by compnerd — APPROVED** [review]

**Comment by compnerd** [comment]
LGTM

**Pull request merged**